preview/clean.js

```diff
@@ -0,0 +1,181 @@
+'use strict';
+
+const child_process = require('child_process');
+const dedent = require('dedent');
+const https = require('https');
+const os = require('os');
+const path = require('path');
+
+module.exports = Cleaner;
+module.exports.exec_git = exec_git;
+
+function Cleaner(token, repo, cache_dir, tmp_dir) {
+  this.repo = repo;
+  let repo_name = path.basename(repo);
+  if (!repo_name.endsWith('.git')) {
+    repo_name += '.git';
+  }
+  this.local_path = `${tmp_dir}/${repo_name}`;
+
+  this.run = () => {
+    return this.clone()
+      .then(this.show_heads)
+      .then(this.heads_to_prs)
+      .then(this.cleanup_closed_prs);
+  }
+
+  this.clone = () => {
+    return exec_git([
+      'clone',
+      '--bare',
+      '--reference', `${cache_dir}/${repo_name}`,
+      repo, this.local_path])
+  }
+
+  this.show_heads = () => {
+    return exec_git(['show-ref', '--heads'], {cwd: this.local_path});
+  }
+
+  this.heads_to_prs = (heads) => {
+    return heads
+      .split('\n')
+      .map(line => {
```

Comment:
Would reduce be better here? I'm assuming the subsequent `filter` is to remove random `undefined`s generated by this map. That way you can conditionally push items on to the accumulator.